### ISSUE TITLE
Python 3.x support 

### ISSUE THREAD

**Issue by breznak** [numenta/nupic#2506]
reference for @oxtopus PR: #2377 where he already did lot of work! 

- [ ] translation with python `2to3` tool
- [ ] use `python -3` to hint incompatible parts for 2to3

**Comment by breznak** [comment]
ups, I've missed https://github.com/numenta/nupic/issues/2378 already

**Comment by rcrowder** [comment]
@breznak No, this issue is good. And ties in with https://github.com/numenta/nupic.core/issues/567 where this work is due to occur first.

**Comment by impredicative** [comment]
This issue has been languishing for years.

**Comment by rhyolight** [comment]
There are folks working on a fork of this repository that will be python 3 compatible at https://discourse.numenta.org/t/python-3-migration/3228/

**Comment by rhyolight** [comment]
@ObadaJabassini see my comment about the fork above.

**Comment by impredicative** [comment]
Any update for an estimated release date for Python 3? Python 2 support and compatibility soon won't be necessary. This migration is holding up adoption of nupic. It is just mind boggling why this migration has to be a community effort and not a core Numenta effort.

**Comment by oxtopus** [comment]
Heard.  Numenta fired everyone working on it and deprioritized anything not
directly related to "research".  If you want Python 3 support, I suggest
looking for alternatives or doing it yourself.

**Comment by kyle-sorensen** [comment]
Lol @ people waiting for Python 3, can't spend 10 seconds using virtualenv to create python 2.7 environment

**Comment by impredicative** [comment]
@kyle-sorensen It is not about creating a virtualenv; that's easy. It's about integrating with other code that's already in Python 3.

More importantly, Python 2 is [about to expire](https://pythonclock.org/) in 1.5 years. I guess I will reach out to Numenta management.

**Comment by rhyolight** [comment]
@impredicative The mission of Numenta is twofold:

1. To understand how intelligence works in the mammalian brain.
1. To implement those systems in non-biological systems

I understand your frustration that NuPIC is out of date, but our priorities put research ahead of software updates for the community at this time. At some point in the future, we will likely cycle back into a software applications implementation mode, at which point it might make sense for us to dedicate engineering resources to migrating to python 3.

* * *

> Wasn't a goal of Jeff H. to let a thousand flowers bloom? Well they're withering right now. If the goal is to work in a silo, I think even Vicarious can do that.

This is not about one implementation of HTM. This is about HTM theory. As I have said many times, I don't think NuPIC is going to be the HTM implementation the takes the world by storm. I want people to understand the concepts behind HTM so they can build their own systems based upon the principles of intelligence we have uncovered.

**Comment by impredicative** [comment]
Numenta has had an open position requiring "experience with Python, and deep learning frameworks such as PyTorch or TensorFlow." Does this mean that there is a plan to rewrite NuPIC using either the PyTorch or TensorFlow ecosystems? If so, that's good. If not, why not? Or is that role for some other goal?

**Comment by dstromberg** [comment]
Has anyone tried nupic (the Numenta version) on a Pypy compatible with 2.7?  I believe the Pypy project intends to keep 2.7 compatibility indefinitely.
